specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,268 @@
+  // Set the list of origins that are allowed to use the scheme.
+  HRESULT SetAllowedOriginsList([in] LPCWSTR* allowedOriginsList, [in] UINT32 allowedOriginsListCount);
+}
+
+// This is the ICoreWebView2StagingCustomSchemeRegistrationCollection Staging interface
+[uuid(6ba6f68e-e7dc-4903-a3cc-6a047b6dcb61), object, pointer_default(unique)]
+interface ICoreWebView2StagingCustomSchemeRegistrationCollection : IUnknown {
+  /// Gets the `ICoreWebView2StagingCustomSchemeRegistration` located in the
+  /// `ICoreWebView2StagingCustomSchemeRegistrationCollection` at the given index.
+  HRESULT GetValueAtIndex([in] UINT32 index,
+                          [out, retval] ICoreWebView2StagingCustomSchemeRegistration** registration);
+
+  /// Inserts a `ICoreWebView2StagingCustomSchemeRegistration` into the
+  /// `ICoreWebView2StagingCustomSchemeRegistrationCollection` at the given index.
+  HRESULT InsertValue(
+      [in] ICoreWebView2StagingCustomSchemeRegistration* value);
+
+  /// The number of `ICoreWebView2StagingCustomSchemeRegistration`s in the collection.
+  [propget] HRESULT Count([out, retval] UINT32* count);
+}
+
+[uuid(ac52d13f-0d38-475a-9dca-876580d6793e), object, pointer_default(unique)]
+interface ICoreWebView2StagingEnvironmentOptions : IUnknown {
+  /// Collection of custom scheme registrations
+  [propget] HRESULT CustomSchemeRegistrar(
+      [out, retval] ICoreWebView2StagingCustomSchemeRegistrationCollection** schemeRegistrar);
+}
+```
+
+## .NET API
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    // Represents the registration of a custom scheme with the CoreWebView2Environment.
+    // This allows the WebView2 app to be able to handle
+    // WebResourceRequested event for requests with the specified scheme and
+    // be able to navigate the WebView2 to the custom scheme. Once the environment
+    // is created, the registrations are valid and immutable throughout the
+    // lifetime of the associated WebView2s' browser process and any WebView2
+    // environments sharing the browser process must be created with identical
+    // custom scheme registrations, otherwise the environment creation will fail.
+    // Any further attempts to register the same scheme will fail.
+    // The URIs of registered custom schemes will be treated similar to http URIs for their origins.
+    // They will have tuple origins for URIs with host and opaque origins for
+    // URIs without host as specified in https://html.spec.whatwg.org/multipage/origin.html
+    // Example:
+    // custom-scheme-with-host://hostname/path/to/resource has origin of custom-scheme-with-host://hostname
+    // custom-scheme-without-host:path/to/resource has origin of custom-scheme-without-host:path/to/resource
+    // For WebResourceRequested event, the request URIs and filter URIs with custom schemes
+    // will be normalized according to RFC3986.
+    runtimeclass CoreWebView2StagingCustomSchemeRegistration
+    {
```

Comment:
You should give this a constructor. The CoreWebView2EnvironmentOptions runtimeclass in the generated IDL should have one if you want to follow its example.